cmd/cluster-replication.go

```diff
@@ -0,0 +1,990 @@
+	for i, v := range arg.Clusters {
+		if i == idx {
+			continue
+		}
+		admClient, err := getAdminClient(v)
+		if err != nil {
+			return fmt.Errorf("unable to create admin client: %w", err)
+		}
+		joinReq.SvcAcctParent = v.AccessKey
+		err = admClient.CRInternalJoin(ctx, joinReq)
+		if err != nil {
+			return fmt.Errorf("error in CR join: %v", err)
+		}
+	}
+
+	state := crState{
+		Name:                    arg.Clusters[idx].Name,
+		Peers:                   joinReq.Peers,
+		ServiceAccountAccessKey: svcCred.AccessKey,
+	}
+	c.Lock()
+	defer c.Unlock()
+	c.state = state
+	c.Enabled = true
+
+	return nil
+}
+
+// InternalJoinReq - internal API handler to respond to a peer cluster's request
+// to join.
+func (c *ClusterReplMgr) InternalJoinReq(ctx context.Context, arg madmin.CRInternalJoinReq) error {
+	if c.Enabled {
+		return errCRCannotJoin
+	}
+
+	var ourName string
+	for d, p := range arg.Peers {
+		if d == globalDeploymentID {
+			ourName = p.Name
+			break
+		}
+	}
+	if ourName == "" {
+		return errCRSelfNotFound
+	}
+
+	_, err := globalIAMSys.NewServiceAccount(ctx, arg.SvcAcctParent, nil, newServiceAccountOpts{
+		accessKey: arg.SvcAcctAccessKey,
+		secretKey: arg.SvcAcctSecretKey,
+	})
+	if err != nil {
+		return err
+	}
+
+	state := crState{
+		Name:                    ourName,
+		Peers:                   arg.Peers,
+		ServiceAccountAccessKey: arg.SvcAcctAccessKey,
+	}
+	return c.saveToDisk(ctx, state)
+}
+
+// GetClusterInfo - returns cluster replication information.
+func (c *ClusterReplMgr) GetClusterInfo(ctx context.Context) (info madmin.ClusterReplicateInfo, err error) {
+	c.RLock()
+	defer c.RUnlock()
+	if !c.Enabled {
+		return info, nil
+	}
+
+	info.Enabled = true
+	info.Name = c.state.Name
+	info.Clusters = make([]madmin.PeerInfo, 0, len(c.state.Peers))
+	for _, peer := range c.state.Peers {
+		info.Clusters = append(info.Clusters, peer)
+	}
+	sort.SliceStable(info.Clusters, func(i, j int) bool {
+		return info.Clusters[i].Name < info.Clusters[j].Name
+	})
+
+	info.ServiceAccountAccessKey = c.state.ServiceAccountAccessKey
+	return info, nil
+}
+
+// MakeBucketHook - called during a regular make bucket call when cluster
+// replication is enabled. It is responsible for the creation of the same bucket
+// on remote clusters, and creating replication rules on local and peer
+// clusters.
+func (c *ClusterReplMgr) MakeBucketHook(ctx context.Context, bucket string, opts BucketOptions) error {
+	// At this point, the local bucket is created.
+
+	c.RLock()
+	defer c.RUnlock()
+	if !c.Enabled {
+		return nil
+	}
+
+	optsMap := make(map[string]string)
+	if opts.Location != "" {
+		optsMap["location"] = opts.Location
+	}
+	if opts.LockEnabled {
+		optsMap["lockEnabled"] = ""
+	}
+	if opts.VersioningEnabled {
+		optsMap["versioningEnabled"] = ""
+	}
+
+	// Create bucket and enable versioning on all peers.
+	for d, p := range c.state.Peers {
+		if d == globalDeploymentID {
+			err := c.PeerBucketMakeWithVerHandler(ctx, bucket, opts)
```

Comment:
Something is confusing me here, you are passing versioningEnabled as option but the function name is called MakeWithVer(sion?), can you elaborate ?

Reply:
You are right - `opts.VersioningEnabled != true` whenever a new bucket is created as there is no way to do that from the PutBucket API afaik. I will fix this part.